kani-driver/src/call_cargo.rs

```diff
@@ -22,66 +22,67 @@ pub struct CargoOutputs {
-}
-
 impl KaniSession {
     /// Calls `cargo_build` to generate `*.symtab.json` files in `target_dir`
     pub fn cargo_build(&self) -> Result<CargoOutputs> {
         let build_target = env!("TARGET"); // see build.rs
-        let target_dir = self.args.target_dir.as_ref().unwrap_or(&find_target_dir()).clone();
+        let metadata = MetadataCommand::new().exec().expect("Failed to get cargo metadata.");
+        let target_dir = self
+            .args
+            .target_dir
+            .as_ref()
+            .unwrap_or(&metadata.target_directory.clone().into())
+            .clone();
         let outdir = target_dir.join(build_target).join("debug/deps");
 
-        let flag_env = {
-            let rustc_args = self.kani_rustc_flags();
-            crate::util::join_osstring(&rustc_args, " ")
-        };
-
-        let mut args: Vec<OsString> = Vec::new();
+        let mut kani_args = self.kani_specific_flags();
+        let mut rustc_args = self.kani_rustc_flags();
 
+        let mut cargo_args: Vec<OsString> = vec!["rustc".into()];
         if self.args.tests {
-            args.push("test".into());
-            args.push("--no-run".into());
-        } else {
-            args.push("build".into());
-        }
-
-        for package in self.args.package.iter() {
-            args.push("--package".into());
-            args.push(package.into());
+            cargo_args.push("--tests".into());
         }
 
         if self.args.all_features {
-            args.push("--all-features".into());
+            cargo_args.push("--all-features".into());
         }
 
-        if self.args.workspace {
-            args.push("--workspace".into());
-        }
-
-        args.push("--target".into());
-        args.push(build_target.into());
+        cargo_args.push("--target".into());
+        cargo_args.push(build_target.into());
 
-        args.push("--target-dir".into());
-        args.push(target_dir.into());
+        cargo_args.push("--target-dir".into());
+        cargo_args.push(target_dir.into());
 
         if self.args.verbose {
-            args.push("-v".into());
+            cargo_args.push("-v".into());
+        }
+
+        // Arguments that will only be passed to the target package.
+        let mut pkg_args: Vec<OsString> = vec![];
+        if self.args.mir_linker {
+            // Only provide reachability flag to the target package.
+            pkg_args.push("--".into());
+            pkg_args.push("--reachability=harnesses".into());
+        } else {
+            // Pass legacy reachability to the target package and its dependencies.
+            kani_args.push("--reachability=legacy".into());
         }
 
-        let mut cmd = Command::new("cargo");
-        cmd.args(args)
-            .env("RUSTC", &self.kani_compiler)
-            .env("RUSTFLAGS", "--kani-flags")
-            .env("KANIFLAGS", flag_env);
+        // Only joing them at the end. All kani flags must come first.
+        kani_args.append(&mut rustc_args);
```

Comment:
&mut? oh, wait this says append not extend. IMO `extend` is cleaner because it avoids needing the mut

Reply:
Sure. I'll use `extend_from_slice()`.